File: src/main/java/me/narutopig/bors/CustomEnchants.java
package me.narutopig.bors;

import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;

import java.lang.reflect.Field;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static me.narutopig.bors.Util.itemStack;

public class CustomEnchants {
    public static final EnchantmentWrapper EXPERIENCE = new EnchantmentWrapper(
            "experience",
            "Experience",
            5,
            itemStack(Material.LAPIS_BLOCK, 1),
            new Enchantment[]{Enchantment.MENDING}
    );

    public static final EnchantmentWrapper AFTERMATH = new EnchantmentWrapper(
            "aftermath",
            "Aftermath",
            1,
            itemStack(Material.TNT, 1)
    );

    public static final EnchantmentWrapper TELEKINESIS = new EnchantmentWrapper(
            "telekinesis",
            "Telekinesis",
            1,
            itemStack(Material.ENDER_PEARL, 16)
    );

    public static final List<EnchantmentWrapper> customEnchants = Arrays.asList(
            AFTERMATH,
            EXPERIENCE,
            TELEKINESIS
    );

    public static void register() {
        List<Boolean> registered = new ArrayList<>();

        for (Enchantment enchantment : customEnchants) {
            registered.add(Arrays.stream(Enchantment.values()).toList().contains(enchantment));
        }

        for (int i = 0; i < customEnchants.size(); i++) {
            if (!registered.get(i)) registerEnchant(customEnchants.get(i));
        }
    }

    public static void registerEnchant(Enchantment enchantment) {
        // definitely not copied

        boolean registered = true;

        try {
            Field f = Enchantment.class.getDeclaredField("acceptingNew");
            f.setAccessible(true);
            f.set(null, true);
            Enchantment.registerEnchantment(enchantment);
        } catch (Exception e) {
            registered = false;
            e.printStackTrace();
        }

        if (registered) {
            System.out.println(enchantment.getName() + " was registered!");
        }
    }
}

File: src/main/kotlin/io/github/narutopig/bors/CustomEnchants.kt
package io.github.narutopig.bors

import io.github.narutopig.bors.util.General
import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import java.util.*
import java.util.stream.Collectors

object CustomEnchants {
    val EXPERIENCE = EnchantmentWrapper(
        "experience",
        "Experience",
        5,
        General.itemStack(Material.LAPIS_BLOCK, 1), arrayOf(Enchantment.MENDING)
    )
    val AFTERMATH = EnchantmentWrapper(
        "aftermath",
        "Aftermath",
        1,
        General.itemStack(Material.TNT, 1)
    )
    val TELEKINESIS = EnchantmentWrapper(
        "telekinesis",
        "Telekinesis",
        1,
        General.itemStack(Material.ENDER_PEARL, 4)
    )
    val customEnchants: MutableList<EnchantmentWrapper> = mutableListOf(
        AFTERMATH,
        EXPERIENCE,
        TELEKINESIS
    )

    fun register() {
        val registered: MutableList<Boolean> = ArrayList()
        for (enchantment in customEnchants) {
            registered.add(Arrays.stream(Enchantment.values()).collect(Collectors.toList()).contains(enchantment))
        }
        for (i in customEnchants.indices) {
            if (!registered[i]) registerEnchant(customEnchants[i])
        }
    }

    private fun registerEnchant(enchantment: Enchantment) {
        // definitely not copied
        var registered = true
        try {
            val f = Enchantment::class.java.getDeclaredField("acceptingNew")
            f.isAccessible = true
            f[null] = true
            Enchantment.registerEnchantment(enchantment)
        } catch (e: Exception) {
            registered = false
            e.printStackTrace()
        }
        if (registered) {
            println(enchantment.key.toString() + " was registered!")
        }
    }

    fun isCustomEnchantment(e: Enchantment): Boolean {
        return customEnchants.any { it.key == e.key }
    }

    fun getCustomEnchant(key: String): EnchantmentWrapper {
        return customEnchants.first { it.key.key == key }
    }
}